### PR TITLE
Test to verify application failover and relocate between managed

### PR DESCRIPTION
**Pull request by Akarsha-rai**
Tests to verify application failover and relocate between managed clusters
Covers two test cases:
 
1) Failover and relocate of only application between managed clusters
2) Failover to secondary cluster when primary cluster is DOWN and Relocate back to primary cluster once it recovers

This test is also compatible to be run from UI, pass the yaml conf/ocsci/dr_ui.yaml to trigger it.

### CODE REVIEW

**Comment by Akarsha-rai** [comment]
Test case passes, fails in teardown in the environment checks to connect to stretched ceph cluster with error: `monclient: keyring not found`

With UI failover and failback of app: 
[test_failover_and_failback_of_app_from_ui.txt](https://github.com/red-hat-storage/ocs-ci/files/11506918/test_failover_and_failback_of_app_from_ui.txt)

With CLI failover and failback of app: 
[test_failover_and_failback_of_app_from_cli.txt](https://github.com/red-hat-storage/ocs-ci/files/11506924/test_failover_and_failback_of_app_from_cli.txt)

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+)
+from ocs_ci.helpers.dr_helpers_ui import (
+    failover_relocate_ui,
+    verify_failover_relocate_status_ui,
+)
+from ocs_ci.utility import version
+
+logger = logging.getLogger(__name__)
+
+polarion_id_primary_up = ""
```
Comment (hnallurv) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:35]: please add polarion id

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+polarion_id_primary_down = "OCS-4346"
+
+
+@tier1
+class TestApplicationFailoverAndRelocate:
+    """
+    Test Failover and Relocate actions for a busybox application
```
Comment (hnallurv) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:42]: please have design such that multiple apps also can be handled.

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        self,
+        setup_acm_ui,
+        primary_cluster_down,
+        nodes_multicluster,
+        dr_workload,
+        node_restart_teardown,
+    ):
+        """
+        Tests to verify application failover and relocate between managed clusters
+        There are two test cases:
```
Comment (hnallurv) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:83]: In future, please have two separate tests for these cases.

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+            1) Failover and relocate of only application between managed clusters
+            2) Failover to secondary cluster when primary cluster is DOWN and Relocate
+                back to primary cluster once it recovers
+
+        This test is also compatible to be run from UI,
+        pass the yaml conf/ocsci/dr_ui.yaml to trigger it.
```
Comment (hnallurv) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:89]: better to have this as part of "usage info"

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        # Verify application are running in other managedcluster
+        # And not in previous cluster
+        set_current_primary_cluster_context(workload.workload_namespace)
+        wait_for_all_resources_creation(
+            workload.workload_pvc_count,
+            workload.workload_pod_count,
+            workload.workload_namespace,
+        )
+
+        # Verify the failover status from UI
```
Comment (hnallurv) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:149]: if this test is run from cli then how are you verifying the failover status?

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        if config.RUN.get("mdr_relocate_via_ui"):
+            config.switch_acm_ctx()
+            verify_failover_relocate_status_ui(acm_obj)
+
+        # Start nodes if cluster is down
+        wait_time = 60
+        if primary_cluster_down:
+            logger.info(
+                f"Waiting for {wait_time} minutes before starting nodes of previous primary cluster"
```
Comment (hnallurv) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:158]: is wait_time in seconds or minutes? If seconds then please change "... minutes before starting...."

File: ocs_ci/helpers/dr_helpers.py [hunk 7]
```diff
@@ -559,14 +559,15 @@ def wait_for_all_resources_deletion(
...
-    logger.info("Waiting for all PVs to be deleted")
-    sample = TimeoutSampler(
-        timeout=timeout,
-        sleep=5,
-        func=get_pv_count,
-        namespace=namespace,
-    )
-    sample.wait_for_func_value(0)
+    if config.MULTICLUSTER["multicluster_mode"] != "metro-dr":
+        logger.info("Waiting for all PVs to be deleted")
```
Comment (hnallurv) [inline ocs_ci/helpers/dr_helpers.py:563]: change this to "Waiting for all PVs to be created"

Comment (Akarsha-rai) [reply ocs_ci/helpers/dr_helpers.py:563]: This function `wait_for_all_resources_deletion`  wait for workload and replication resources to be deleted, in this step we are waiting for all PVs are deleted.

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 2 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:42]: In `mdr_workload.yaml` we need to pass the repo details, where we can pass such that multiple applications to be created.

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 6 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:158]: oops, it is in seconds. Will change accordingly

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+        # Verify resources creation on preferredCluster
+        set_current_primary_cluster_context(workload.workload_namespace)
+        wait_for_all_resources_creation(
+            workload.workload_pvc_count,
+            workload.workload_pod_count,
+            workload.workload_namespace,
+        )
+
+        # Verify Relocate statis from UI
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:217]: will take up in next PR

**Review by raghavendra-talur — APPROVED** [review]

File: ocs_ci/helpers/dr_helpers_ui.py [hunk 9]
```diff
@@ -251,13 +254,9 @@ def failover_relocate_ui(
...
         else:
             log.info("Select target cluster on ACM UI")
-            acm_obj.do_click(
-                format_locator(
-                    acm_loc["failover-preferred-cluster-name"],
-                    failover_or_preferred_cluster,
-                ),
-                enable_screenshot=True,
-            )
+            xpath = f"//button[text()='{failover_or_preferred_cluster}']"
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers_ui.py:257]: Is this change needed?

File: ocs_ci/ocs/acm/acm.py [hunk 10]
```diff
@@ -366,6 +366,8 @@ def login_to_acm():
         driver (Selenium WebDriver)
 
     """
+
+    config.switch_acm_ctx()
```
Comment (am-agrawa) [inline ocs_ci/ocs/acm/acm.py:370]: Context change may not be required here

File: ocs_ci/ocs/dr/dr_workload.py [hunk 11]
```diff
@@ -120,6 +121,7 @@ class BusyBox(DRWorkload):
 
     def __init__(self, **kwargs):
         workload_repo_url = config.ENV_DATA["dr_workload_repo_url"]
+        log.info(f"{workload_repo_url}")
```
Comment (am-agrawa) [inline ocs_ci/ocs/dr/dr_workload.py:124]: Better to add a text before printing URL

File: ocs_ci/ocs/dr/dr_workload.py [hunk 12]
```diff
@@ -238,11 +244,16 @@ def delete_workload(self, force=False):
             log.info("Verify backend RBD images are deleted")
             for cluster in get_non_acm_cluster_config():
                 config.switch_ctx(cluster.MULTICLUSTER["multicluster_index"])
+                rbd_pool_name = (
+                    (config.ENV_DATA.get("rbd_name") or rbd_name)
+                    if config.DEPLOYMENT["external_mode"]
+                    else constants.DEFAULT_CEPHBLOCKPOOL
```
Comment (am-agrawa) [inline ocs_ci/ocs/dr/dr_workload.py:250]:
```suggestion
                    else rbd_pool_name
```

But why replacing `constants.DEFAULT_CEPHBLOCKPOOL` with `rbd_pool_name` is needed overall?

File: ocs_ci/ocs/node.py [hunk 13]
```diff
@@ -2693,3 +2693,24 @@ def generate_nodes_for_provider_worker_node_tests():
...
+    from ocs_ci.ocs import platform_nodes
+
+    node_objs = get_node_objs()
+    factory = platform_nodes.PlatformNodesFactory()
+    nodes = factory.get_nodes_platform()
+    waiting_time = 30
+    for node in node_objs:
+        node_name = node.name
+        unschedule_nodes([node_name])
+        drain_nodes([node_name])
```
Comment (am-agrawa) [inline ocs_ci/ocs/node.py:2711]: Is it gracefull drain?

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 2 above)
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:42]: The doctext says `busybox` only but the use case should be generic IMO

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 4 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:89]: +1

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 3 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:83]: +1

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        set_current_primary_cluster_context(workload.workload_namespace)
+        primary_cluster_index = config.cur_index
+        node_objs = get_node_objs()
+        primary_cluster_name = get_current_primary_cluster_name(
+            namespace=workload.workload_namespace
+        )
+        self.drcluster_name = primary_cluster_name
+        self.namespace = workload.workload_namespace
+
+        # Stop primary cluster nodes
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:113]: Shouldn't we run IOs for some time before starting failover?

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+        # Create application on Primary managed cluster
+        set_current_primary_cluster_context(workload.workload_namespace)
+        primary_cluster_index = config.cur_index
+        node_objs = get_node_objs()
+        primary_cluster_name = get_current_primary_cluster_name(
+            namespace=workload.workload_namespace
+        )
+        self.drcluster_name = primary_cluster_name
+        self.namespace = workload.workload_namespace
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:111]: I don't think we need another variable assignment here

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+                logger.error(
+                    "ODF/ACM version isn't supported for Failover/Relocate operation"
+                )
+                raise NotImplementedError
+
+        acm_obj = AcmAddClusters()
+        workload = dr_workload(num_of_subscription=1)[0]
+
+        # Create application on Primary managed cluster
+        set_current_primary_cluster_context(workload.workload_namespace)
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:104]: Current primary context can only be set after workloads are deployed, isn't it?

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+        primary_cluster_name = get_current_primary_cluster_name(
+            namespace=workload.workload_namespace
+        )
+        self.drcluster_name = primary_cluster_name
+        self.namespace = workload.workload_namespace
+
+        # Stop primary cluster nodes
+        if primary_cluster_down:
+            logger.info("Stopping primary cluster nodes")
+            nodes_multicluster[primary_cluster_index].stop_nodes(node_objs)
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:116]: After nodes are down, don't we want to check the cluster status on ACM console using function `check_cluster_status_on_acm_console` in case Failover is done via UI?

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+        # Verify resources creation on preferredCluster
+        set_current_primary_cluster_context(workload.workload_namespace)
+        wait_for_all_resources_creation(
+            workload.workload_pvc_count,
+            workload.workload_pod_count,
+            workload.workload_namespace,
+        )
+
+        # Verify Relocate statis from UI
+        if config.RUN.get("rdr_relocate_via_ui"):
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:218]:
```suggestion
        if config.RUN.get("mdr_relocate_via_ui"):
```

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+            config.switch_acm_ctx()
+            verify_failover_relocate_status_ui(acm_obj)
+
+        # Start nodes if cluster is down
+        wait_time = 60
+        if primary_cluster_down:
+            logger.info(
+                f"Waiting for {wait_time} minutes before starting nodes of previous primary cluster"
+            )
+            time.sleep(60)
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:160]:
```suggestion
            time.sleep(wait_time)
```

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+            nodes_multicluster[primary_cluster_index].start_nodes(node_objs)
+            logger.info(
+                "Waiting for some minutes after starting nodes of previous primary cluster"
+            )
+            time.sleep(120)
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:165]:
```suggestion
            time.sleep(2*wait_time)
```

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+        validate_data_integrity(workload.workload_namespace)
+
+        # Unfenced the managed cluster which was Fenced earlier
+        enable_unfence(drcluster_name=self.drcluster_name)
+
+        # Reboot the nodes which unfenced
+        gracefully_reboot_ocp_nodes(workload.workload_namespace, self.drcluster_name)
+
+        # Application Relocate to Primary managed cluster
+        secondary_cluster_name = get_current_secondary_cluster_name(
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:189]: Run IOs for some time before relocate

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 8 above)
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:217]:
```suggestion
        # Verify Relocate status from UI
```

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+            nodes_multicluster[primary_cluster_index].stop_nodes(node_objs)
+
+        # Fenced the primary managed cluster
+        enable_fence(drcluster_name=self.drcluster_name)
+
+        # Application Failover to Secondary managed cluster
+        secondary_cluster_name = get_current_secondary_cluster_name(
+            workload.workload_namespace
+        )
+        if config.RUN.get("mdr_failover_via_ui"):
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:125]: When failover is done via UI, UI validation is missing for the same using function `verify_failover_relocate_status_ui`

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+        setup_acm_ui,
+        primary_cluster_down,
+        nodes_multicluster,
+        dr_workload,
+        node_restart_teardown,
+    ):
+        """
+        Tests to verify application failover and relocate between managed clusters
+        There are two test cases:
+            1) Failover and relocate of only application between managed clusters
```
Comment (am-agrawa) [inline tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:84]:
```suggestion
            1) Failover and relocate of application between managed clusters
```

**Review by parikshithb — APPROVED** [review]

File: ocs_ci/ocs/acm/acm.py (hunk 10 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/acm/acm.py]: deleted

**Comment by Akarsha-rai** [comment]
[test_failover_and_failback_of_app_from_ui.txt](https://github.com/red-hat-storage/ocs-ci/files/11815957/test_failover_and_failback_of_app_from_ui.txt)

**Review by parikshithb — DISMISSED** [review]

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 4 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:90]: Will take in next PR

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+            2) Failover to secondary cluster when primary cluster is DOWN and Relocate
+                back to primary cluster once it recovers
+
+        This test is also compatible to be run from UI,
+        pass the yaml conf/ocsci/dr_ui.yaml to trigger it.
+        """
+
+        if config.RUN.get("mdr_failover_via_ui"):
+            ocs_version = version.get_semantic_ocs_version_from_config()
+            if ocs_version <= version.VERSION_4_12:
```
[the remark this answers is not in the document]
Comment (hnallurv) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:95]: Looks like this is not required at the moment to be done

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 5 above)
Comment (hnallurv) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:159]: Please open a github issue for tracking

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hnallurv) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py]: Please open a github issue for tracking this

File: ocs_ci/helpers/dr_helpers.py (hunk 7 above)
Comment (hnallurv) [reply ocs_ci/helpers/dr_helpers.py:563]: my bad. this is not an issue.

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py]: raised an issue to track https://github.com/red-hat-storage/ocs-ci/issues/7885, along with this there additional validation needs to be added

**Review by hnallurv — DISMISSED** [review]

File: ocs_ci/helpers/dr_helpers.py [hunk 25]
```diff
@@ -606,3 +607,105 @@ def get_all_drpolicy():
...
+    config.switch_ctx(restore_index)
+
+
+def get_fence_state(drcluster_name):
+    """
+    Returns the clusterfence state of given drcluster
+    Args:
+        drcluster_name (str): Name of the DRcluster
+    Returns:
+        state (str): If drcluster are fenced: Fenced or Unfenced, else None if not defined
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply ocs_ci/helpers/dr_helpers.py:689]: done

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 15 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:112]: Pending

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 14 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:114]: Pending

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 17 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:117]: Pending

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 22 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:135]: Pls highlight where?

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 21 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:199]: Is it done?

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 15 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py]: sorry, I missed this. Removed self.drcluster variable as said using existing variable `primary_cluster_name`

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 14 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:116]: as discussed on call, in MDR we don't have VR, so don't need to wait till the sync happens once Pod is created it should run IOs (using dd)

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 17 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:119]: this been added below so marking them resolved

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 22 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:137]: It was added but there was typo in conditional, corrected to `mdr_failover_via_ui`

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 20 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py]: done, so resoling it

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 21 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:203]: As discussed in call, we don't have VR concept in MDR so additional wait for sync is not required. So marking them resolved

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 8 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py]: changed

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 2 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py]: done

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 18 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py]: done

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 19 above)
Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py]: done

File: ocs_ci/ocs/dr/dr_workload.py (hunk 12 above)
Comment (am-agrawa) [reply ocs_ci/ocs/dr/dr_workload.py:247]: ack

File: ocs_ci/ocs/node.py (hunk 13 above)
Comment (am-agrawa) [reply ocs_ci/ocs/node.py:2713]: discussed over call, looks fine as per MDR use case

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 16 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:109]: ack

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 14 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:116]: ack

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 17 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:119]: now looks fine

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 21 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:203]: ack

File: ocs_ci/helpers/dr_helpers_ui.py [hunk 26]
```diff
@@ -238,6 +240,7 @@ def failover_relocate_ui(
             format_locator(acm_loc["select-policy"], policy_name),
             enable_screenshot=True,
         )
+
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers_ui.py:243]: Extra line

File: tests/disaster-recovery/mdr/test_app_failover_and_relocate.py (hunk 17 above)
Comment (am-agrawa) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:119]: The same check `check_cluster_status_on_acm_console` is missing when failover is performed by UI but primary cluster is up and running. @Akarsha-rai

Comment (Akarsha-rai) [reply tests/disaster-recovery/mdr/test_app_failover_and_relocate.py:119]: added

File: ocs_ci/helpers/dr_helpers_ui.py (hunk 26 above)
Comment (Akarsha-rai) [reply ocs_ci/helpers/dr_helpers_ui.py]: removed

File: ocs_ci/helpers/dr_helpers_ui.py [hunk 27]
```diff
@@ -195,7 +195,9 @@ def failover_relocate_ui(
     Function to perform Failover/Relocate operations via ACM UI
 
     Args:
+
         acm_obj (AcmAddClusters): ACM Page Navigator Class
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers_ui.py:199]: Extra line added again

**Review by am-agrawa — APPROVED** [review]
Great work :+1:

File: ocs_ci/helpers/dr_helpers_ui.py (hunk 27 above)
Comment (Akarsha-rai) [reply ocs_ci/helpers/dr_helpers_ui.py:198]: removed

**Review by hnallurv — APPROVED** [review]

**Comment by Akarsha-rai** [comment]
[Uploading test_failover_and_failback_of_app_from_ui.txt…]()

**Review by am-agrawa — APPROVED** [review]

**Review by parikshithb — APPROVED** [review]

File: ocs_ci/ocs/node.py [hunk 28]
```diff
@@ -2693,3 +2693,26 @@ def generate_nodes_for_provider_worker_node_tests():
...
     return generated_nodes
+
+
+def gracefully_reboot_nodes():
+    """
+
+    Gracefully reboot OpenShift Container Platform nodes
+
+    """
+    from ocs_ci.ocs import platform_nodes
```
Comment (PrasadDesala) [inline ocs_ci/ocs/node.py:2704]: pls move the import to the top of the file

File: conf/ocsci/mdr_workload.yaml [hunk 29]
```diff
@@ -0,0 +1,6 @@
+ENV_DATA:
+  dr_workload_repo_url: "https://github.com/Akarsha-rai/ocs-workloads.git"
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply conf/ocsci/mdr_workload.yaml:2]: yes, that's plan which is tracked by issue [7885](https://github.com/red-hat-storage/ocs-ci/issues/7885).

File: ocs_ci/helpers/dr_helpers.py [hunk 30]
```diff
@@ -606,3 +609,112 @@ def get_all_drpolicy():
     drpolicy_obj = ocp.OCP(kind=constants.DRPOLICY)
     drpolicy_list = drpolicy_obj.get(all_namespaces=True).get("items")
     return drpolicy_list
+
+
+def enable_fence(drcluster_name):
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply ocs_ci/helpers/dr_helpers.py:614]: Will take this in next PR, issue is tracked here https://github.com/red-hat-storage/ocs-ci/issues/7885.

File: ocs_ci/ocs/node.py (hunk 28 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/node.py:2704]: Was facing cyclic import error, so to avoid it imported inside the function

**Review by PrasadDesala — APPROVED** [review]
